diff --git a/Sources/CodeEditSourceEditor/Controller/TextViewController+Lifecycle.swift b/Sources/CodeEditSourceEditor/Controller/TextViewController+Lifecycle.swift
index bb654a4..d5b70dd 100644
--- a/Sources/CodeEditSourceEditor/Controller/TextViewController+Lifecycle.swift
+++ b/Sources/CodeEditSourceEditor/Controller/TextViewController+Lifecycle.swift
@@ -14,10 +14,16 @@ extension TextViewController {
         // The calculation this causes cannot be done until the view knows it's final position
         updateTextInsets()
         minimapView.layout()
+        highlighter?.refreshVisibleRanges()
     }

     override public func viewDidAppear() {
         super.viewDidAppear()
+        textView.updateFrameIfNeeded()
+        highlighter?.refreshVisibleRanges()
+        DispatchQueue.main.async { [weak self] in
+            self?.highlighter?.refreshVisibleRanges()
+        }
         textCoordinators.forEach { $0.val?.controllerDidAppear(controller: self) }
     }

diff --git a/Sources/CodeEditSourceEditor/Enums/CaptureName.swift b/Sources/CodeEditSourceEditor/Enums/CaptureName.swift
index 32b37aa..e8b643f 100644
--- a/Sources/CodeEditSourceEditor/Enums/CaptureName.swift
+++ b/Sources/CodeEditSourceEditor/Enums/CaptureName.swift
@@ -54,7 +54,7 @@ public enum CaptureName: Int8, CaseIterable, Sendable {
             return .include
         case "constructor":
             return .constructor
-        case "keyword":
+        case "keyword", "operator", "keyword.operator", "type.qualifier", "label", "punctuation.special":
             return .keyword
         case "boolean":
             return .boolean
@@ -66,27 +66,27 @@ public enum CaptureName: Int8, CaseIterable, Sendable {
             return .tag
         case "comment":
             return .comment
-        case "variable":
+        case "variable", "constant":
             return .variable
-        case "property":
+        case "property", "property.definition", "variable.member":
             return .property
-        case "function":
+        case "function", "function.call", "function.builtin", "function.macro", "function.macro.builtin":
             return .function
-        case "method":
+        case "method", "method.call":
             return .method
         case "number":
             return .number
         case "float":
             return .float
-        case "string":
+        case "string", "string.regex", "string.special", "string.escape":
             return .string
-        case "type":
+        case "type", "type.builtin":
             return .type
-        case "parameter":
+        case "parameter", "variable.parameter":
             return .parameter
-        case "type_alternate":
+        case "type_alternate", "type.definition":
             return .typeAlternate
-        case "variable.builtin":
+        case "variable.builtin", "constant.builtin":
             return .variableBuiltin
         case "keyword.return":
             return .keywordReturn
diff --git a/Sources/CodeEditSourceEditor/Highlighting/HighlightProviding/HighlightProviderState.swift b/Sources/CodeEditSourceEditor/Highlighting/HighlightProviding/HighlightProviderState.swift
index c4ae800..5252323 100644
--- a/Sources/CodeEditSourceEditor/Highlighting/HighlightProviding/HighlightProviderState.swift
+++ b/Sources/CodeEditSourceEditor/Highlighting/HighlightProviding/HighlightProviderState.swift
@@ -91,6 +91,7 @@ class HighlightProviderState {
         self.visibleRangeProvider = visibleRangeProvider

         highlightProvider.setUp(textView: textView, codeLanguage: language)
+        invalidate()
     }

     func setLanguage(language: CodeLanguage) {
diff --git a/Sources/CodeEditSourceEditor/Highlighting/Highlighter.swift b/Sources/CodeEditSourceEditor/Highlighting/Highlighter.swift
index cefd82b..50a27f6 100644
--- a/Sources/CodeEditSourceEditor/Highlighting/Highlighter.swift
+++ b/Sources/CodeEditSourceEditor/Highlighting/Highlighter.swift
@@ -115,6 +115,7 @@ class Highlighter: NSObject {
                 language: language
             )
         }
+        visibleRangeProvider.visibleTextChanged()
     }

     // MARK: - Public
@@ -128,6 +129,10 @@ class Highlighter: NSObject {
         highlightProviders.forEach { $0.invalidate(set) }
     }

+    public func refreshVisibleRanges() {
+        visibleRangeProvider.visibleTextChanged()
+    }
+
     /// Sets the language and causes a re-highlight of the entire text.
     /// - Parameter language: The language to update to.
     public func setLanguage(language: CodeLanguage) {
diff --git a/Sources/CodeEditSourceEditor/Highlighting/StyledRangeContainer/StyledRangeContainer.swift b/Sources/CodeEditSourceEditor/Highlighting/StyledRangeContainer/StyledRangeContainer.swift
index c77e2b5..70bb41b 100644
--- a/Sources/CodeEditSourceEditor/Highlighting/StyledRangeContainer/StyledRangeContainer.swift
+++ b/Sources/CodeEditSourceEditor/Highlighting/StyledRangeContainer/StyledRangeContainer.swift
@@ -140,7 +140,6 @@ extension StyledRangeContainer: HighlightProviderStateDelegate {
     func applyHighlightResult(provider: ProviderID, highlights: [HighlightRange], rangeToHighlight: NSRange) {
         assert(rangeToHighlight != .notFound, "NSNotFound is an invalid highlight range")
         guard var storage = _storage[provider]?.store else {
-            assertionFailure("No storage found for the given provider: \(provider)")
             return
         }
         var runs: [RangeStoreRun<StyleElement>] = []
diff --git a/Sources/CodeEditSourceEditor/Theme/EditorTheme.swift b/Sources/CodeEditSourceEditor/Theme/EditorTheme.swift
index c44cfc9..ee6c667 100644
--- a/Sources/CodeEditSourceEditor/Theme/EditorTheme.swift
+++ b/Sources/CodeEditSourceEditor/Theme/EditorTheme.swift
@@ -83,12 +83,13 @@ public struct EditorTheme: Equatable {
     /// - Returns: Theme attributes for the capture.
     private func mapCapture(_ capture: CaptureName?) -> Attribute {
         switch capture {
-        case .include, .constructor, .keyword, .boolean, .variableBuiltin,
-                .keywordReturn, .keywordFunction, .repeat, .conditional, .tag:
+        case .include, .constructor, .keyword, .keywordReturn, .keywordFunction, .repeat, .conditional, .tag:
             return keywords
+        case .boolean, .variableBuiltin:
+            return values
         case .comment: return comments
         case .variable, .property: return variables
-        case .function, .method: return variables
+        case .function, .method: return commands
         case .number, .float: return numbers
         case .string: return strings
         case .type: return types
diff --git a/Sources/CodeEditSourceEditor/TreeSitter/TreeSitterClient.swift b/Sources/CodeEditSourceEditor/TreeSitter/TreeSitterClient.swift
index 44a0a34..d1da632 100644
--- a/Sources/CodeEditSourceEditor/TreeSitter/TreeSitterClient.swift
+++ b/Sources/CodeEditSourceEditor/TreeSitter/TreeSitterClient.swift
@@ -129,7 +129,7 @@ public final class TreeSitterClient: HighlightProviding {
         }

         executor.cancelAll(below: .all)
-        if forceSyncOperation {
+        if forceSyncOperation || textView.documentRange.length <= Constants.maxSyncContentLength {
             executor.execSync(operation)
         } else {
             executor.execAsync(priority: .reset, operation: operation, onCancel: {})
